app/features/build_runner/remote_runner.rb

```diff
@@ -20,187 +28,193 @@ class RemoteRunner
     # The commit sha we want to run the build for
     attr_reader :sha
 
-    # All lines that were generated so far, this might not be a complete run
-    # This is an array of hashes
-    attr_accessor :all_build_output_log_rows
-
-    # TODO: Add comment
-    attr_accessor :build_change_observer_blocks
-
     # TODO: extract this thing out of this class.
     attr_reader :github_service
 
-    def initialize(project:, git_fork_config:, trigger:, github_service:)
+    # TODO(snatchev): consider how to reduce the number of dependencies to this class.
+    def initialize(project:, git_fork_config:, trigger:, github_service:, grpc: Agent::Client.new("localhost"))
       @project = project
       @git_fork_config = git_fork_config
-      @all_build_output_log_rows = []
-      @build_change_observer_blocks = []
       @github_service = github_service
-
       @current_build = prepare_build_object(trigger: trigger)
-      # This build needs to be saved during initialization because RemoteRunner.start
-      # is called by a TaskQueue and could potentially happen after the BuildController tries to
-      # fetch this build.
-      save_build_status_locally!
-
-      @artifacts_path = Dir.mktmpdir
-
-      fastlane_log_path = File.join(@artifacts_path, "fastlane.log")
-      @fastlane_log = Logger.new(fastlane_log_path)
-      @current_build.artifacts << Artifact.new(
-        type: "fastlane.log",
-        reference: fastlane_log_path,
-        provider: @project.artifact_provider
-      )
+      @grpc = grpc
 
-      @build_state = MicroMachine.new(:PENDING).tap do |fsm|
```

Comment:
Removing the state machine here. I feel pretty strongly that the only state should be defined in the Agent and it should be the source of truth about the state of the build.

The RemoteRunner will accept any state transition and pass it along to the subscribers.